### PR TITLE
build.sh: simplify and semi-modernize

### PR DESCRIPTION
**Pull request by aktau**
1. All POSIX shells support $(...) syntax [1], including /bin/sh. shellcheck
   warns about it.
2. [[ won't work in /bin/sh [2], so use [ everywhere. I wonder why it worked
   now, perhaps the test was running on a system where /bin/sh -> /bin/bash.
3. In POSIX sh, string indexing is undefined. [SC2039]. Unfortunately, this
   means we require a subprocess: https://wiki.ubuntu.com/DashAsBinSh. Very
   ugly.
4. In POSIX sh, arrays are undefined. We seem to use $flags as an array
   after constructing it by string concatenation. I tried to verify that
   this has the same effect as just passing the quoted string in bash:

      bash-3.2$ flags="-DCMAKE_GOOK"
      bash-3.2$ flags="$flags -DCMAKE_MOARMA"
      bash-3.2$ flags="$flags -DCMAKE_URURURUR"
      bash-3.2$ ./argv $flags
      0: ./argv
      1: -DCMAKE_GOOK
      2: -DCMAKE_MOARMA
      3: -DCMAKE_URURURUR
      bash-3.2$ ./argv "${flags[@]}"
      0: ./argv
      1: -DCMAKE_GOOK -DCMAKE_MOARMA -DCMAKE_URURURUR
      bash-3.2$ ./argv "$flags"
      0: ./argv
      1: -DCMAKE_GOOK -DCMAKE_MOARMA -DCMAKE_URURURUR
      bash-3.2$

5. Enable exit on unknown variable (-u). All variables should be known,
   otherwise we have an error in the script. shellcheck doesn't warn so I
   think it's fine.

Apart from shellcheck(1), I also ran checkbashisms(1). The latter only
reported that "command -v" might not be available in other shells.
Apparently only ash(1) doesn't understand it.

NOTE: Why are we even trying to support pre-Mavericks (Darwin < 13) OSX? We
don't even support the most modern OSX (El Capitan) fully yet. OSX upgrades
are free and generally don't leave old machines behind. Most machines made
after 2009 can upgrade to El Capitan, AFAIK. I also believe that systems
that have all the utilities and libraries necessary to build PCSX2 will have
/bin/bash >= 3.x.

NOTE 2: Does cmake/ninja generate the same type of output in
compile_commands.json?

[1]: http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_03
[2]: http://serverfault.com/a/52050

### CODE REVIEW

**Comment by mirh** [comment]
Are you going to still keep #297 open?

And then I dunno if OSX is more like linux (newer == better) or more like Windows (old can be better), but I guess you'll know if pre-maverick is more like XP or more like 7.

**Comment by aktau** [comment]
> Are you going to still keep #297 open?

I was under the impression the majority of those changes were first imported by @juhalaukkanen and then merged via one of their PRs. There's no real need to keep exactly that one open, although I dream of resuming that work at one point. I remember getting stuck at some assembly that wasn't allowed on x86-64 (but was allowed on x86).

> And then I dunno if OSX is more like linux (newer == better) or more like Windows (old can be better), but I guess you'll know if pre-maverick is more like XP or more like 7 to keep supported.

It's a UNIX, so more like Linux. The point is that OSX was never supported by PCSX2, so it's now easy to say that only versions 10.11+ will be supported.

**Comment by gregory38** [comment]
on 1/ why not? but why is is better ? `$(..)` needs more char and harder to highligh properly in editor.

Script used to be sh. But then I got convinced by bad people that bash was much better. I put back sh recently  because some users got issue to use the file. I didn't use any tool to validate it so some bad stuffs remain (such as 2/ and 4/)

Good question on ninja json format.

File: build.sh [hunk 1]
```diff
@@ -225,17 +225,16 @@ if [ "$clangTidy" -eq 1 ] && [ `command -v clang-tidy` ]; then
         #clang-tidy -p $compile_json $cpp -checks='*' -header-filter='.*'
 
         # Don't check header, don't check google/llvm coding conventions
-        echo "$count/$total"
```
Comment (gregory38) [inline build.sh]: It would be nice to fix the code. Or add a progress bar, I don't know if there is any standard stuff.

**Comment by gregory38** [comment]
Did you test command -v for command that don't exist. Just to validate the syntax without `[ ]`

**Comment by aktau** [comment]
Yes I did, in /bin/sh :)

File: build.sh (hunk 1 above)
Comment (gregory38) [inline build.sh]: Or a better solution will be to multithread it a bit. Half an hour to get result is damn slow.

**Comment by gregory38** [comment]
Ok perfect.

**Comment by mirh** [comment]
@aktau Yes, I imagined one day you'd have liked to make it advance, but I was wondering whether that got definitively surpassed (and new work would basically require all new commits) or if on the other hand you wasn't keeping it open because there's still something worth.

File: build.sh (hunk 1 above)
Comment (aktau) [inline build.sh]: I agree. There's an easy solution that doesn't require a lot (or any) synchronization: we can use GNU parallel (xargs would parallelize but also mangle the output). I'll submit a PR.

**Comment by aktau** [comment]
> @aktau Yes, I imagined one day you'd have liked to make it advance, but I was wondering whether that got definitively surpassed (and new work would basically require all new commits) or if on the other hand you wasn't keeping it open because there's still something worth.

I'm not sure if _all_ of the commits I had got merged into @juhalaukkanen's branch, and it's difficult to see at a glance because they squashed those commits into bigger ones. I guess I'll notice whenever I try to rebase my branch. In any case, the most important work was already merged.

**Pull request merged**